package com.rest.services.test;

import io.restassured.response.Response;

import org.apache.log4j.Logger;
import org.testng.Assert;
import org.testng.annotations.Test;
import org.testng.asserts.SoftAssert;

import utils.ReusableMethods;
import utils.TestBase;

import com.google.gson.Gson;
import com.rest.responsepojo.CreateJiraIssueResponse;
import com.rest.responsepojo.LoginResponse;
import com.rest.services.Service;

/**
 * creating jira api tests {@link #loginToJiraApiTest()}
 * {@link #createIssueInJiraApiTest()}
 * 
 * @author TEChinnmaya41
 *
 */

public class JiraTests extends TestBase {

	SoftAssert softAssert;
	private static Logger log = Logger.getLogger(JiraTests.class.getName());

	@Test
	public void loginToJiraApiTest() {

		service = new Service();
		try {
			log.info("Test case loginToJiraApiTest started");
			Response res = service.jiraLoginAuth(
					ReusableMethods.getPropData("username"),
					ReusableMethods.getPropData("password"));

			Gson gson = new Gson();
			LoginResponse loginRes = gson.fromJson(res.asString(),
					LoginResponse.class);

			String sessionId = res.getSessionId();

			softAssert = new SoftAssert();
			softAssert.assertEquals(res.getStatusCode(), 200);
			softAssert.assertEquals(res.getContentType(),
					"application/json;charset=UTF-8");
			softAssert
					.assertEquals(sessionId, loginRes.getSession().getValue());
			log.info("sessionId generated during login:  "+sessionId);
			softAssert.assertAll();
		} catch (Throwable e) {

			log.error("test case failed due to exception--" + e.getMessage());
			Assert.fail();
		}
		log.info("Test case loginToJiraApiTest ends");
	}

	@Test
	public void createIssueInJiraApiTest() throws Exception {

		service = new Service();
		try {
			Response res = service.createIssueInJira(
					ReusableMethods.getPropData("key"),
					ReusableMethods.getPropData("summary"),
					ReusableMethods.getPropData("description"),
					ReusableMethods.getPropData("name"));
			Gson gson = new Gson();
			CreateJiraIssueResponse createJiraIssueResponse = gson.fromJson(
					res.asString(), CreateJiraIssueResponse.class);
			softAssert = new SoftAssert();
			softAssert.assertEquals(res.getStatusCode(), 201);
			log.info("ID Generated by issue creation: "+createJiraIssueResponse.getId());
			log.info("Key generated by issue creation: "+createJiraIssueResponse.getKey());
			log.info("Self generated by Issue creation: "+createJiraIssueResponse.getSelf());
			softAssert.assertAll();

		} catch (Throwable e) {
			log.error("Test case failed due to the exception" + e.getMessage());
			Assert.fail("create issue in Jira failed due to",e);
		}
	}

}
